### PR TITLE
Add tag system to properties

### PR DESCRIPTION
**Pull request by MFraters**
This add a system where features can be tagged, and the dominant tag indices can be requested. I still need to give access to the tag names to outside applications and update all the tests.

### CODE REVIEW

**Comment by MFraters** [comment]
This is now ready for review.

**Comment by gassmoeller** [comment]
@tjhei can you take a look? I wasnt part of the discussion about this.

File: include/world_builder/features/feature_utilities.h [hunk 1]
```diff
@@ -84,6 +85,15 @@ namespace WorldBuilder
         // The local thickness of the segment at the location of the plane.
         double local_thickness;
       };
+
+
+      /**
+       * Add a string to a vector of strings, if the exact string isn't areadly
```
Comment (tjhei) [inline include/world_builder/features/feature_utilities.h:91]:
```suggestion
       * Add a string to a vector of strings, if the exact string isn't already 
```

File: include/world_builder/features/interface.h [hunk 2]
```diff
@@ -130,6 +130,12 @@ namespace WorldBuilder
          */
         std::string name;
 
+        /**
+         * The index of the tag for this feature.
+         * This corresponds to the index in the feature_tags variable which is store in the world.
```
Comment (tjhei) [inline include/world_builder/features/interface.h:135]:
```suggestion
         * This corresponds to the index in the feature_tags variable which is store in the World.
```

File: include/world_builder/world.h [hunk 3]
```diff
@@ -120,15 +120,19 @@ namespace WorldBuilder
...
+       * Grains are identified by 3. The second entry is the grain composition number and the third
        * entry is the number of grains. A query about the grains, where it asks about composition 1
-       * (for example enstatite) and 500 grains, looks like this: {2,1,500}.
+       * (for example enstatite) and 500 grains, looks like this: {3,1,500}.
        * A composition query prodoces n_grains*10 entries in the output vector. The first n_grains
        * entries are the sizes of all the grains, and the other 9 entries are sets of rotation
        * matrices. The rotation matrix entries are ordered [0][0],[0][1],[0][2],[1][0],[1][1],etc.
+       * 
+       * The tag is identified by 4 and no extra information is needed. So the tag
+       * input usually looks like {1,0,0}. A tag query prodoces one entry in the output
```
Comment (tjhei) [inline include/world_builder/world.h:134]:
```suggestion
       * input usually looks like {4,0,0}. A tag query prodoces one entry in the output
```
?

File: include/world_builder/world.h [hunk 4]
```diff
@@ -120,15 +120,19 @@ namespace WorldBuilder
...
+       * vector, represeting the tag of the last/dominant feature.
```
Comment (tjhei) [inline include/world_builder/world.h:135]:
```suggestion
       * vector, representing the index of the tag of the last/dominant feature.
```

File: source/world_builder/features/interface.cc [hunk 5]
```diff
@@ -82,7 +82,12 @@ namespace WorldBuilder
                 prm.declare_entry("model", Types::String("",it.first),
                                   "The name which the user has given to the feature.");
                 prm.declare_entry("name", Types::String(""),
-                                  "The name which the user has given to the feature.");
+                                  "The name which the user has given to the feature. "
+                                  "This is mostly used for documentation purposes, and should in most cases be unique, "
+                                  "although this is not enforced.");
+                prm.declare_entry("tag", Types::String(""),
+                                  "A tag which can be given to a feature. This is meant to catagorize different features. "
```
Comment (tjhei) [inline source/world_builder/features/interface.cc:89]:
```suggestion
                                  "A tag which can be given to a feature. This is meant to categorize different features. "
```

**Review by tjhei — APPROVED** [review]
This looks good. I think we need to document tags a bit better in a few places (maybe once we use them in gwb-grid)

**Review by tjhei — APPROVED** [review]

**Pull request merged**